src/rgw/rgw_common.h

```diff
@@ -2079,6 +2079,27 @@ static inline void buf_to_hex(const unsigned char* const buf,
   }
 }
 
+/* A variant of buf_to_hex dedicated to handling the run-time bufferlist
+ * which can't be hexed in compile-time. It's optimized to avoid extra
+ * linearisation of the incoming buffers. */
+static inline std::string buf_to_hex(const ceph::bufferlist& bl)
+{
+  std::string hexed;
+  hexed.resize(bl.length() * 2);
+
+  size_t consumed = 0;
+  for (const auto buf : bl.buffers()) {
```

Comment:
copying the `buffer::ptr` here induces an atomic increment and decrement of `buffer::raw::nref`, so it's better to take these by reference

Reply:
Thanks for pointing this out, @cbodley! Reworked.